sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala

```diff
@@ -167,36 +167,39 @@ object InMemoryFileIndex extends Logging {
       hadoopConf: Configuration,
       filter: PathFilter,
       sparkSession: SparkSession): Seq[(Path, Seq[FileStatus])] = {
+    // Filter out the directory before listing it leaf files
+    val filteredPaths = paths.filter(filter.accept(_))
 
     // Short-circuits parallel listing when serial listing is likely to be faster.
-    if (paths.size <= sparkSession.sessionState.conf.parallelPartitionDiscoveryThreshold) {
-      return paths.map { path =>
+    if (filteredPaths.size <= sparkSession.sessionState.conf.parallelPartitionDiscoveryThreshold) {
```

Comment:
My concern is whether the filter is supposed to be applied to the original paths? it seems like it's a filter to be applied to what's inside the paths. I'm not clear that's correct, or if it generally is, whether it would ever have an effect in practice? would you write a filter that probably matched your root dirs but also its contents? They're typically things like filtering for extensions or prefixes of the contents

Reply:
PathFilter should be applied to the all hdfs input path from the doc. So I think it should be work for the directory as well.

Reply:
I see, right, it's really a filter on the full path so it doesn't matter when you check. It won't do anything if you're filtering for `.endsWith(".tmp")` or something and listing `/my/path`. It would change behavior, I guess, if you happened to try to list directory `/my/path.tmp` in this case. Now the whole thing would be filtered out, whereas the intent (I think) and current behavior is to only match files underneath.

That's why I wonder what the use case is here.

Reply:
I understand your concern. How about we add some debug log to info the filtered directory or files.  Because the PathFilter should be working on a path, people shouldn't suppose it just works on the name of the current path.

Reply:
No, the PathFilter is applied on a path, not on the leaf file paths. The path can be file or directory. > Is the use case just setting a filter to apply an operation to a subset of dirs?

just as I described in the pr or in the UT test.

> why not filter them out at the caller level?

Configuration should be more flexible. Otherwise, people may need to filter in the user code. Maybe we can add another option in the datafram reader if we must filter by the user code directly.

Reply:
Right now, it's only applied to the full path of what is listed by the 'leaf file' methods. I'm not sure what you mean here by leaf file path then. The point is to _also_ apply it to the path of parent dirs earlier, right?

I think the use case could be made clearer. I'm guessing it's something like this:

You want to read a subset of all `hdfs://name:port/root-dir/timestamp=yyyy-MM-dd/xxx.csv` files for example. You only want to read, say, `hdfs://name:port/root-dir/timestamp=2019-MM-dd/xxx.csv` files really. You pass a filter that matches something like `*2019-MM-dd*` and use all the `hdfs://name:port/root-dir/timestamp=yyyy-MM-dd` directories as the data paths. It has to list all the `hdfs://name:port/root-dir/timestamp=yyyy-MM-dd` paths before it can figure out that lots are filtered out and never had to be listed.

If that's the case, that seems a lot easier to filter in user code. Is the use case different or more complex?

I'm trying to weigh the problem with the behavior change against a realistic use case that would benefit.

Reply:
Hi @srowen, it's OK to filter the path in user code for me. However, it can trigger a Spark job to filter the subdirectory when the directory need be processed is really large by the current implementation. It would be very slowly in user code.

Reply:
Why would anything need a Spark job here... if you're just taking a set of 100 dirs and filtering them according to a predicate before passing to Spark?

Reply:
I just mention the benefit of the current implementation, right? The directory can have a list of subdirectories. I still don't think the PathFilter just applied on a leaf file. I can close it if you really think this is useless.